lib/cc/resolv_override.rb

```diff
@@ -0,0 +1,46 @@
+require "resolv-replace"
+
+module CC
+  # Allows for control of how DNS is resolved so that when we resolve a host for
+  # the purposes of checking that it's not internal, we can ensure that future
+  # resolutions (done as part of making the request) will find the same IP
+  # address.
+  module ResolvOverride
+    def self.get_and_fix_address(host)
+      dns.getaddress(host).tap do |address|
+        fixed.add_address(host, address)
+      end
+    end
+
+    def self.dns
+      @dns ||= Resolv::DNS.new
+    end
+
+    def self.fixed
+      @fixed ||= ResolvOverride::FixedAddress.new
+    end
+
+    class FixedAddress < Resolv::DNS
+      def add_address(host, address)
+        addresses[host] = address
+      end
+
+      def each_address(name, &proc)
+        if addresses.key?(name)
+          yield addresses.fetch(name)
+        end
+      end
+
+      private
+
+      def addresses
+        @addresses ||= {}
+      end
+    end
+  end
+end
+
+Resolv::DefaultResolver.replace_resolvers([
```

Comment:
Downside would be that if this gem is added to any project (and it's OSS, so possible), and is required (either explicitly or via auto-require) that this behavior may take effect? (Even if someone is just adding this gem to e.g. use one specific behavior)

Safer would be to expose a class method like `CC::Services.override_dns_resolution` or similar and then call that from _outside_ the gem if you want that behavior.